src/plan/barriers.rs

```diff
@@ -30,69 +32,100 @@ impl Barrier for NoBarrier {
     fn post_write_barrier(&mut self, _target: WriteTarget) {}
 }
 
-#[derive(Default)]
-pub struct ModBuffer {
-    modified_nodes: Vec<ObjectReference>,
-    modified_edges: Vec<Address>,
-}
-
-pub struct FieldRememberingBarrier<E: ProcessEdgesWork, S: Space<E::VM>> {
+pub struct ObjectRememberingBarrier<E: ProcessEdgesWork, S: Space<E::VM>> {
     mmtk: &'static MMTK<E::VM>,
-    nursery: &'static S,
-    mod_buffer: ModBuffer,
+    _nursery: &'static S,
+    modbuf: Vec<ObjectReference>,
+    meta: SideMetadataSpec,
 }
 
-impl<E: ProcessEdgesWork, S: Space<E::VM>> FieldRememberingBarrier<E, S> {
+impl<E: ProcessEdgesWork, S: Space<E::VM>> ObjectRememberingBarrier<E, S> {
     #[allow(unused)]
-    pub fn new(mmtk: &'static MMTK<E::VM>, nursery: &'static S) -> Self {
+    pub fn new(mmtk: &'static MMTK<E::VM>, nursery: &'static S, meta: SideMetadataSpec) -> Self {
```

Comment:
Though you pass in a constant for `SideMetadataSpec`, it probably is not treated as a constant when you use the field `meta`. Does it matter for performance?

Reply:
I guess it is possible to treat them as constants if we pass them as a set of associated constants wrapped in a generic type parameter. I'll fix this (and the other one).

Reply:
I was just pointing out this could potentially be an issue. If in your measurement, this is not an issue then we dont need to do anything for it.